File: okta-direct-auth-java-cli-sample/src/main/java/com/okta/directauth/cli/model/AuthMethod.java
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.directauth.cli.model;

import com.okta.directauth.model.OobChannel;
import com.okta.directauth.model.PrimaryFactor;

/**
 * Supported authentication methods for the CLI sample app.
 *
 * <p>Each method maps to a display label for CLI menus and can produce the corresponding SDK {@link
 * PrimaryFactor} for authentication requests.
 */
public enum AuthMethod {
  PASSWORD("Password"),
  OTP("OTP"),
  SMS("SMS"),
  VOICE("Voice"),
  OKTA_VERIFY("Push (Okta Verify)");

  private final String label;

  AuthMethod(String label) {
    this.label = label;
  }

  /**
   * Returns the display label for CLI menus.
   *
   * @return the human-readable label
   */
  public String getLabel() {
    return label;
  }

  /**
   * Converts this authentication method to the corresponding SDK {@link PrimaryFactor}.
   *
   * @param passCode the passcode or password entered by the user (used by PASSWORD and OTP)
   * @return the SDK factor for this authentication method
   */
  public PrimaryFactor asFactor(String passCode) {
    if (this == PASSWORD) {
      return new PrimaryFactor.Password(passCode);
    } else if (this == OTP) {
      return new PrimaryFactor.Otp(passCode);
    } else if (this == SMS) {
      return new PrimaryFactor.Oob(OobChannel.SMS);
    } else if (this == VOICE) {
      return new PrimaryFactor.Oob(OobChannel.VOICE);
    } else if (this == OKTA_VERIFY) {
      return new PrimaryFactor.Oob(OobChannel.PUSH);
    }
    throw new IllegalStateException("Unknown auth method: " + this);
  }
}

File: okta-direct-auth-shared/src/commonMain/kotlin/com/okta/directauth/app/model/AuthMethod.kt
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.directauth.app.model

import com.okta.directauth.model.OobChannel
import com.okta.directauth.model.PrimaryFactor
import com.okta.directauth.model.SecondaryFactor

sealed class AuthMethod(
    val label: String,
) {
    open fun asFactor(passCode: String = ""): PrimaryFactor =
        when (this) {
            Password -> PrimaryFactor.Password(passCode)
            Mfa.Otp -> PrimaryFactor.Otp(passCode)
            Mfa.OktaVerify -> PrimaryFactor.Oob(OobChannel.PUSH)
            Mfa.Sms -> PrimaryFactor.Oob(OobChannel.SMS)
            Mfa.Voice -> PrimaryFactor.Oob(OobChannel.VOICE)
            Mfa.Passkeys -> PrimaryFactor.WebAuthn
        }

    // MFA capable Auth methods
    sealed class Mfa(
        label: String,
    ) : AuthMethod(label) {
        override fun asFactor(passCode: String): SecondaryFactor =
            when (this) {
                Otp -> PrimaryFactor.Otp(passCode)
                OktaVerify -> PrimaryFactor.Oob(OobChannel.PUSH)
                Sms -> PrimaryFactor.Oob(OobChannel.SMS)
                Voice -> PrimaryFactor.Oob(OobChannel.VOICE)
                Passkeys -> PrimaryFactor.WebAuthn
            }

        data object Otp : Mfa("OTP")

        data object OktaVerify : Mfa("Okta Verify")

        data object Sms : Mfa("SMS")

        data object Voice : Mfa("Voice")

        data object Passkeys : Mfa("Passkeys")
    }

    data object Password : AuthMethod("Password")

    companion object {
        /**
         * Internal mapping function from string label to AuthMethod.
         * Returns null for unknown labels.
         */
        internal fun stringToAuthMethodInternal(label: String): AuthMethod? =
            when (label) {
                "Password" -> Password
                "OTP" -> Mfa.Otp
                "Okta Verify" -> Mfa.OktaVerify
                "SMS" -> Mfa.Sms
                "Voice" -> Mfa.Voice
                "Passkeys" -> Mfa.Passkeys
                else -> null
            }

        /**
         * Converts a string label to an AuthMethod.
         * Throws IllegalArgumentException if the label is unknown.
         */
        fun fromString(label: String): AuthMethod =
            stringToAuthMethodInternal(label)
                ?: throw IllegalArgumentException("Unknown AuthMethod label: $label")
    }
}

/**
 * Extension function to convert a string to an AuthMethod.
 * Returns null for unknown labels.
 */
fun String.toAuthMethod(): AuthMethod? = AuthMethod.stringToAuthMethodInternal(this)
